---
title: "_VariantExperiment_: A RangedSummarizedExperiment container for VCF/GDS data with GDS back-end"
author: "Qian Liu, Martin Morgan"
- name: Qian Liu
  affiliation: Roswell Park Comprehensive Cancer Center, Buffalo, NY
- name: Hervé Pagès
  affiliation: Fred Hutchinson Cancer Research Center, Seattle, WA
- name: Martin Morgan
  affiliation: Roswell Park Comprehensive Cancer Center, Buffalo, NY
date: "last edit: 07/02/2018"
output:
    BiocStyle::html_document:
        toc: true
        toc_float: true
package: GDSArray
vignette: |
    %\VignetteIndexEntry{VariantExperiment: A RangedSummarizedExperiment container for VCF/GDS data with GDS back-end}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r options, eval=TRUE, echo=FALSE}
options(showHeadLines=3)
options(showTailLines=3)
```

# Introduction 

[VariantExperiment][] is a Bioconductor package for saving the VCF/GDS
format data into RangedSummarizedExperiment object. The
high-throughput genetic/genomic data are saved in [GDSArray][] objects
with GDS back-end. The annotation data for variants/genes/samples are
saved in [DelayedDataFrame][] format with mono-dimensional GDSArray in
each column. The interface of RangedSummarizedExperiment-like data
format enables easy and common manipulations for high-throughput
genetic/genomic data and achieves on-disk reading and processing with
common [SummarizedExperiment][] operation in _R_ and _Bioconductor_.

[VariantExperiment]: https://bioconductor.org/packages/VariantExperiment 
[GDSArray]: https://bioconductor.org/packages/GDSArray
[SummarizedExperiment]: https://bioconductor.org/packages/SummarizedExperiment

# Installation

Users should download the package from Bioconductor.

```{r getPackage, eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("VariantExperiment")
```
The development version is also available to download through github. 
```{r getDevel, eval=FALSE}
devtools::install_github("Bioconductor/VariantExperiment")
```
Load the package into R session.
```{r Load, message=FALSE}
library(VariantExperiment)
```

# Background

## GDSArray

[GDSArray][] is a _Bioconductor_ package that represents GDS files as
objects derived from the [DelayedArray][] package and `DelayedArray`
class. It converts a GDS node in the file into a
`DelayedArray`-derived data structure. The rich common methods and
data operations defined on `GDSArray` makes it more _R_-user-friendly
than working with the GDS file directly. The array data from GDS files
are always returned with the first dimension being features (e.g.,
`genes/variants/snps`) and the second dimension being `samples`. This
feature is consistent with the assay data saved in
`SummarizedExperiment`, and makes the `GDSArray` package interoperable
with other established _Bioconductor_ data
infrastructure.   
More details about GDS or GDSArray format can be found in the
vignettes of the [gdsfmt][], [SNPRelate][], [SeqArray][], [GDSArray][]
and [DelayedArrayy][] packages.

[DelayedArray]: https://bioconductor.org/packages/DelayedArray
[gdsfmt]: https://bioconductor.org/packages/gdsfmt
[SNPRelate]: https://bioconductor.org/packages/SNPRelate
[SeqArray]: https://bioconductor.org/packages/SeqArray

The `GDSArray()` constructor takes 2 arguments: the file path and the
GDS node name inside the GDS file. 

```{r, GDSArray}
file <- SeqArray::seqExampleFileName("gds")
GDSArray(file, "genotype/data")
GDSArray(file, "sample.id")
```

## DelayedDataFrame
[DelayedDataFrame][] is a _Bioconductor_ package that implements
delayed operations on DataFrame objects using standard DataFrame
metaphor. Each column of data inside `DelayedDataFrame` is represented
as `GDSArray` with on-disk GDS file. The slot of `lazyIndex` saves the
mapping indexes of `GDSArray` data in each column for easy
manipulations. Methods like `show`, `validity check`, `[`, `[[`
subsetting, `rbind`, `cbind` are implemented for DelayedDataFrame to
be operated around lazyIndex. The listData slot in `DelayedDataFrame`
stays untouched until a realization call like `DataFrame()`
constructor or `as.list()` incurred.
[DelayedDataFrame][https://github.com/Bioconductor/DelayedDataFrame]

# VariantExperiment

The package of `VariantExperiment` is implemented to represent VCF/GDS
files using standard SummarizedExperiment metaphor. It is a container
for high-through genetic/genomic data with GDS back-end, and is
interoperable with the statistical functions/methods that are
implemented in `SNPRelate`, `SeqArray` and `SeqVarTools` that are
designed for GDS data. The `SummarizedExperiment` metaphor also gets
the benefit of common manipulations within _Bioconductor_ ecosystem
that are more user-friendly.

## VariantExperiment class and coercion methods

`VariantExperiment` class is defined to inherit
`RangedSummarizedExperiment`. The difference would be that the assay
data are saved as `GDSArray`, and the annotation data are saved by
default as `DelayedDataFrame` (with option to save as ordinary
`DataFrame`), both of which are representing the data on-disk with GDS
back-end.

### Coercion function from `VCF` to `VariantExperiment`

The coercion function of `VCF2VE` could convert the `vcf` file
directly into `VariantExperiment` object. To achieve the best storage
efficiency, the assay data are saved in `GDSArray` format, and the
annotation data are saved in `DelayedDataFrame` format (with no option
of ordinary `DataFrame`), which could be retrieved by `rowData()` for
feature related annotations and `colData()` for sample related
annotations (Only when `sample.info` argument is specified).

```{r VCF2VE}
vcf <- SeqArray::seqExampleFileName("vcf")
ve <- VCF2VE(vcf)
ve
assay(ve, 1)
rowData(ve)
``` 

User could also have the opportunity to save the sample related
annotation info directly into the `VariantExperiment` object.

```{r sampleInfo}
sampleInfo <- system.file("extdata", "Example_sampleInfo.txt",
                          package="VariantExperiment")
ve <- VCF2VE(vcf, sample.info = sampleInfo)
colData(ve)
```

Arguments could be specified to take specific info columns or format
columns from the vcf file. The `start` and `count` arguments could be
used to specify the start position and number of variants to read into
`Variantexperiment` object.

```{r VCF2VEArgs}
ve1 <- VCF2VE(vcf, info.import=c("OR", "GP"))
rowData(ve1)
ve2 <- VCF2VE(vcf, start=101, count=1000)
dim(ve2)
```

### Coercion function from `GDS` to `VariantExperiment`

```{r GDS2VE}
gds <- SeqArray::seqExampleFileName("gds")
ve <- makeSummarizedExperimentFromGDS(gds)
ve
```
## VariantExperiment-methods (VCF2VE, makeSummarizedExperimentfromGDS, titv... vcf related.)
## common manipulations... (subsetting, save new VE file, load VE file, ... )


## SummarizedExperiment with GDS back-end assay data.

GDSArray is the GDS back-end DelayedArray, which could stick into the SummarizedExperiment by function `makeSummarizedExperimentFromGDS()`. 
```{r, SEGDSArray}
se <- makeSummarizedExperimentFromGDS(file)
se
names(assays(se))
assay(se, 1)
```

By default, the `rowData` and `colData` are also read into the SummarizedExperiment object in `GDSArray` format. User could use the arguments `rowDataOnDisk` and `colDataOnDisk` to change the setting.
```{r, rowColData}
rowData(se)
colData(se)
se1 <- makeSummarizedExperimentFromGDS(file, rowDataOnDisk=FALSE, colDataOnDisk=FALSE)
rowData(se1)
colData(se1)
```
User could also use the function `showAvailable` for available values for other arguments in the function of `makeSummarizedExperimentFromGDS`. Note that the `infoColumns` for `SEQ_ARRAY` data will save as columns inside the `rowData(SE)`, with the prefix of "info_".
```{r, showAvail}
args(makeSummarizedExperimentFromGDS)
showAvailable(file)
se2 <- makeSummarizedExperimentFromGDS(file, rowDataColumns=c("ID", "REF", "ALT"), infoColumns=c("AA", "DP", "OR", "GP"), rowDataOnDisk=FALSE)
rowData(se2)
rowRanges(se2)
```

## `save/loadGDSSummarizedExperiment`
Use the function `saveGDSSummarizedExperiment` to save any SE object (with ordinary in-memory / GDSArray-based assay data) into the GDSArray-based SE object. Arguments of `rowDataOnDisk` and `colDataOnDisk` are supplied for users to save the `rowData(se) / colData(se)` with GDSArray-based `DataFrame`. By default, all the data slots of `SummarizedExperiment` objects (`assays()`, `rowData()`, `colData()`) are saved with GDSArray-based data.

```{r, save}
## save se1 as ordinary in-memory for all data slots
file <- SeqArray::seqExampleFileName("gds")
se <- makeSummarizedExperimentFromGDS(file, rowDataOnDisk=F, colDataOnDisk=F)
se1 <- se
for(i in seq_along(assays(se1))){
    assays(se1)[[i]] <- unname(as.array(assays(se1)[[i]]))
}
lapply(assays(se1), class)
rowData(se1)
colData(se1)

## saveGDSSummarizedExperiment
## drty <- tempdir()
se2 <- saveGDSSummarizedExperiment(se1, replace=T, fileFormat="SEQ_ARRAY")
se2
lapply(assays(se2), class)
assays(se2)[[1]]
rowData(se2)
colData(se2)
```

```{r, object_size}
## global parameter for showing less rows in each R output
options()
se0 <- makeSummarizedExperimentFromGDS(file)
print(object.size(se0), units="auto")  ## assays on-disk, row/colData on-disk
print(object.size(se), units="auto")  ## assays on-disk, row/colData in-memory
print(object.size(se0[TRUE, TRUE]), units="auto")
print(object.size(se[TRUE, TRUE]), units="auto")

print(object.size(assays(se)), units="auto")
## 280.7 Kb
print(object.size(se1), units="auto") ## assays/row/colData in-memory
print(object.size(assays(se1)), units="auto")
print(object.size(assays(se1, withDimnames=FALSE)), units="auto")
## 2.1 Mb
print(object.size(se2), units="auto")  ## row/colData on-disk
print(object.size(assays(se2)), units="auto")
```

Use the function `loadGDSSummarizedExperiment` to load the newly generated GDSArray-based SE object. 
```{r, load}
se3 <- loadGDSSummarizedExperiment()
se3
all.equal(assays(se2), assays(se3))
rm(se3)
```

## VariantExperiment common SE operations
### Getters and setters
- `rowRanges()` / (`rowData()`), `colData()`, `metadata()`
```{r getSet}
library(SummarizedExperiment)
rowRanges(se2)
head(colData(se2))
head(rowData(se2))
```

### subsetting functions
- two-dimensional subsetting
```{r, 2d}
se2[1:10, 1:5]
```
- `$` operates on `colData()` columns, for easy sample extraction.
```{r colDataExtraction}
head(colData(se2))
se2[, as.logical(se2$family == "1328")]   ## FIXME
```
- subsetting by `rowData()` columns.
```{r rowDataExtraction}
head(rowData(se2))
se2[as.logical(rowData(se2)$REF == "T"),]   ## FIXME
```

### Range-based operations

- `subsetByOverlaps()`
`SummarizedExperiment` objects support all of the `findOverlaps()` methods and associated functions.  This includes `subsetByOverlaps()`, which makes it easy to subset a `SummarizedExperiment` object by an interval.

```{r overlap}
subsetByOverlaps(se2, GRanges("22:1-48958933"))
assay(subsetByOverlaps(se2, GRanges("22:1-48958933")))
rowData(subsetByOverlaps(se2, GRanges("22:1-48958933")))
```


# Future work
- Make the meta data for rows and cols on-disk reading.  -- done!
- More interoperable with other GDS based functions, including DNA-sequencing data annotation, filteration, and statistical analysis.   -- done!
- Develop a pipeline for using `VariantExperiment` as the basic data structure for DNA-sequencing data analysis.  -- next version. 

# Session Info
```{r sessionInfo}
sessionInfo()
```
