---
title: "_VariantExperiment_: A Ranged Summarized Experiment container for GDS back end data"
author: "Qian Liu, Martin Morgan"
date: "Revised: 12 Jan, 2017"
abstract: >
  
  VariantExperiment is a Bioconductor package for saving the GDS format data into SummarizedExperiment object. The high-throughput genetic/genomic data saved in GDS format is converted into GDSArray object, which is an GDS back-end DelayedArray, and then stick into SummarizedExperiment to achieve on-disk reading and processing of array data with common SummarizedExperiment operation. 

output:
  BiocStyle::html_document:
    toc: true
Vignette: >
  %\VignetteIndexEntry{VariantExperiment}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r style, echo=FALSE, results='asis'}
BiocStyle::markdown()
```

# Introduction 

```{r getPackage, eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("VariantExperiment")
```

```{r Load}
library(VariantExperiment)
library(gdsfmt)
```

## GDS format
The package `gdsfmt` provides a high-level R interface to CoreArray Genomic Data Structure (GDS) data files, it has hierarchical structure and could store multiple scalable array-oriented data sets with metadata information. It is designed for large-scale datasets, especially for data which are much larger than the available random-access memory. The GDS format has been widely used in genetic/genomic research for high-throughput genotyping or sequencing data. There are two major classes that extends the `gds.class`: `SNPGDSFileClass` suited for genotyping data (e.g., GWAS), and `SeqVarGDSClass` that are designed specifically for DNA-sequencing data. The file format attribute in each data classes are set as `SNP_ARRAY` and `SEQ_ARRAY`. There are rich functions written based on these data classes for common data operation and statistical analysis. 

### SNP_ARRAY 
`SNP_ARRAY` file format is defined in the package of `SNPRelate`, in the class of `SNPGDSFileClass`. This gds data format is designed for the snp array data, e.g., GWAS. 

```{r, snpArray}
file <- system.file("extdata", "hapmap_geno.gds", package = "SNPRelate")
SNPRelate::snpgdsSummary(file)
f <- SNPRelate::snpgdsOpen(file)
f
### required gds nodes.
ls.gdsn(f)
### required attributes for SNPGDSFileClass. 
get.attr.gdsn(f$root)
get.attr.gdsn(index.gdsn(f, "genotype"))
SNPRelate::snpgdsClose(f)
```
### SEQ_ARRAY
`SEQ_ARRAY` file format is defined in the package of `SeqArray`, in the class of `SeqVarGDSClass`.
```{r, seqArray}
file <- system.file(package="VariantExperiment", "extdata", "CEU_Exon.gds")
f <- SeqArray::seqOpen(file)
ls.gdsn(f)
### required gds nodes: sample.id, variant.id, position, chromosome, allele, and genotype.
get.attr.gdsn(index.gdsn(f, "description"))
node <- index.gdsn(f, "genotype/data")
objdesp.gdsn(node)$dim
### [1]    2   90 1348   (ploidy x sample x variants)
aperm(readex.gdsn(node, sel=list(1:2, 1:3, 1:5)), perm=c(3,2,1))
SeqArray::seqClose(f)
```

# VariantExperiment
The package of `VariantExperiment` is implemented to represent GDS using standard SummarizedExperiment metaphor. It could save high-throughput genetic/genomic data (e.g., genotyping/sequencing data) within the `SummarizedExperiment` structure. It is more like a container right now. We are working to make it more interoperable with the existing functions in `SNPRelate` and `SeqVarTools` designed for gds data. 

## GDSArraySeed, GDSArray, and DelayedArray

GDSArray is the GDS back-end data for DelayedArray, which could be easily stick into SummarizedExperiment, and have on-disk reading and processing of the assay data in SE. Common operations on `SummarizedExperiment` could be applied to the assays in GDSArray format, including the subsetting, getters and setters.  
```{r, GDSArray}
seed <- GDSArraySeed(file, "genotype/data")
slotNames(seed)
```
If the slot `permut` equals to `TRUE`, it means that the gdsnode inside the gds file is in a permuted dimension as in the `GDSArraySeed` and `GDSArray`. The first two dimensions of `GDSArraySeed` and `GDSArray` always corresponds to `genes/variants/snps` and `samples`. When using `[` for subsetting, it will read the on-disk gds data in a permuted way.   
```{r, permute}
seed@permute
ga <- GDSArray(file)  
ga <- GDSArray(seed)  ### equivalent
ga[1:5, 1:3, ]  ## The genotype for first 5 variants, and first 3 samples.
f <- openfn.gds(file)
aperm(readex.gdsn(index.gdsn(f, "genotype/data"), sel=list(1:2, 1:3, 1:5)), perm=c(3,2,1))
closefn.gds(f)
```
The `DelayedArray(GDSArraySeed)` will return the `GDSArray` data, which has the same content as the `GDSMatrix` returned from `GDSArray(GDSArraySeed)`.
```{r, da}
da <- DelayedArray(seed)
class(da)
```

## SummarizedExperiment with GDS back-end assay data.

GDSArray is the GDS back-end DelayedArray, which could stick into the SummarizedExperiment by function `makeSummarizedExperimentFromGDS()`. 
```{r, SEGDSArray}
se <- makeSummarizedExperimentFromGDS(file)
se
names(assays(se))
assay(se, 1)
```

By default, the `rowData` and `colData` are also read into the SummarizedExperiment object in `GDSArray` format. User could use the arguments `rowDataOnDisk` and `colDataOnDisk` to change the setting.
```{r, rowColData}
rowData(se)
colData(se)
se1 <- makeSummarizedExperimentFromGDS(file, rowDataOnDisk=FALSE, colDataOnDisk=FALSE)
rowData(se1)
colData(se1)
```
User could also use the function `showAvailable` for available values for other arguments in the function of `makeSummarizedExperimentFromGDS`. Note that the `infoColumns` for `SEQ_ARRAY` data will save as columns inside the `rowData(SE)`, with the prefix of "info_".
```{r, showAvail}
args(makeSummarizedExperimentFromGDS)
showAvailable(file)
se2 <- makeSummarizedExperimentFromGDS(file, rowDataColumns=c("ID", "REF", "ALT"), infoColumns=c("AA", "DP", "OR", "GP"), rowDataOnDisk=FALSE)
rowData(se2)
rowRanges(se2)
```


## `save/loadGDSSummarizedExperiment`
Use the function `saveGDSSummarizedExperiment` to save any SE object (with ordinary in-memory / GDSArray-based assay data) into the GDSArray-based SE object. Arguments of `rowDataOnDisk` and `colDataOnDisk` are supplied for users to save the `rowData(se) / colData(se)` with GDSArray-based `DataFrame`. By default, all the data slots of `SummarizedExperiment` objects (`assays()`, `rowData()`, `colData()`) are saved with GDSArray-based data.

```{r, save}
## save se1 as ordinary in-memory for all data slots
file <- SeqArray::seqExampleFileName("gds")
se <- makeSummarizedExperimentFromGDS(file, rowDataOnDisk=F, colDataOnDisk=F)
se1 <- se
for(i in seq_along(assays(se1))){
    assays(se1)[[i]] <- unname(as.array(assays(se1)[[i]]))
}
lapply(assays(se1), class)
rowData(se1)
colData(se1)

## saveGDSSummarizedExperiment
## drty <- tempdir()
se2 <- saveGDSSummarizedExperiment(se1, replace=T, fileFormat="SEQ_ARRAY")
se2
lapply(assays(se2), class)
assays(se2)[[1]]
rowData(se2)
colData(se2)
```

```{r, object_size}
print(object.size(se[TRUE, TRUE]), units="auto")
print(object.size(se), units="auto")  ## assays on-disk, row/colData in-memory
print(object.size(se1), units="auto") ## assays/row/colData in-memory
print(object.size(assays(se)), units="auto")
## 280.7 Kb
print(object.size(assays(se1)), units="auto")
## 2.1 Mb
print(object.size(se2), units="auto")  ## row/colData on-disk
print(object.size(assays(se2)), units="auto")
```

Use the function `loadGDSSummarizedExperiment` to load the newly generated GDSArray-based SE object. 
```{r, load}
se3 <- loadGDSSummarizedExperiment()
se3
all.equal(assays(se2), assays(se3))
rm(se3)
```

## VariantExperiment common SE operations
### Getters and setters
- `rowRanges()` / (`rowData()`), `colData()`, `metadata()`
```{r getSet}
library(SummarizedExperiment)
rowRanges(se2)
head(colData(se2))
head(rowData(se2))
```

### subsetting functions
- two-dimensional subsetting
```{r, 2d}
se2[1:10, 1:5]
```
- `$` operates on `colData()` columns, for easy sample extraction.
```{r colDataExtraction}
head(colData(se2))
se2[, as.logical(se2$family == "1328")]   ## FIXME
```
- subsetting by `rowData()` columns.
```{r rowDataExtraction}
head(rowData(se2))
se2[as.logical(rowData(se2)$REF == "T"),]   ## FIXME
```

### Range-based operations

- `subsetByOverlaps()`
`SummarizedExperiment` objects support all of the `findOverlaps()` methods and associated functions.  This includes `subsetByOverlaps()`, which makes it easy to subset a `SummarizedExperiment` object by an interval.

```{r overlap}
subsetByOverlaps(se2, GRanges("22:1-48958933"))
assay(subsetByOverlaps(se2, GRanges("22:1-48958933")))
rowData(subsetByOverlaps(se2, GRanges("22:1-48958933")))
```


# Future work
- Make the meta data for rows and cols on-disk reading.  -- done!
- More interoperable with other GDS based functions, including DNA-sequencing data annotation, filteration, and statistical analysis.   -- next sprint!
- Develop a pipeline for using `VariantExperiment` as the basic data structure for DNA-sequencing data analysis.  

# Session Info
```{r sessionInfo}
sessionInfo()
```
