---
title: "_VariantExperiment_: A Ranged Summarized Experiment container for GDS back end data"
author: "Qian Liu, Martin Morgan"
date: "Revised: 22 Nov, 2017"
abstract: >
  
  VariantExperiment is a Bioconductor package for saving the GDS format data into SummarizedExperiment object. The high-throughput genetic/genomic data saved in GDS format is converted into GDSArray object, which is an GDS back-end DelayedArray, and then stick into SummarizedExperiment to achieve on-disk reading and processing of array data with common SummarizedExperiment operation. 

output:
  BiocStyle::html_document:
    toc: true
Vignette: >
  %\VignetteIndexEntry{VariantExperiment}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r style, echo=FALSE, results='asis'}
BiocStyle::markdown()
```

# Introduction 

```{r getPackage, eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("VariantExperiment")
```

```{r Load}
library(VariantExperiment)
library(gdsfmt)
```

## GDS format
The package `gdsfmt` provides a high-level R interface to CoreArray Genomic Data Structure (GDS) data files, it has hierarchical structure and could store multiple scalable array-oriented data sets with metadata information. It is designed for large-scale datasets, especially for data which are much larger than the available random-access memory. The GDS format has been widely used in genetic/genomic research for high-throughput genotyping or sequencing data. There are two major classes that extends the `gds.class`: `SNPGDSFileClass` suited for genotyping data (e.g., GWAS), and `SeqVarGDSClass` that are designed specifically for DNA-sequencing data. The file format attribute in each data classes are set as `SNP_ARRAY` and `SEQ_ARRAY`. There are rich functions written based on these data classes for common data operation and statistical analysis. 

### SNP_ARRAY 
`SNP_ARRAY` file format is defined in the package of `SNPRelate`, in the class of `SNPGDSFileClass`. This gds data format is designed for the snp array data, e.g., GWAS. 

```{r, snpArray}
file <- system.file("extdata", "hapmap_geno.gds", package = "SNPRelate")
SNPRelate::snpgdsSummary(file)
f <- SNPRelate::snpgdsOpen(file)
f
### required gds nodes.
ls.gdsn(f)
### required attributes for SNPGDSFileClass. 
get.attr.gdsn(f$root)
get.attr.gdsn(index.gdsn(f, "genotype"))
SNPRelate::snpgdsClose(f)
```
### SEQ_ARRAY
`SEQ_ARRAY` file format is defined in the package of `SeqArray`, in the class of `SeqVarGDSClass`.
```{r, seqArray}
file <- system.file(package="VariantExperiment", "extdata", "CEU_Exon.gds")
f <- SeqArray::seqOpen(file)
ls.gdsn(f)
### required gds nodes: sample.id, variant.id, position, chromosome, allele, and genotype.
get.attr.gdsn(index.gdsn(f, "description"))
node <- index.gdsn(f, "genotype/data")
objdesp.gdsn(node)$dim
### [1]    2   90 1348   (ploidy x sample x variants)
readex.gdsn(node, sel=list(1:2, 1:3, 1:3))
seqArray::seqClose(f)
```

# VariantExperiment
The package of `VariantExperiment` is implemented to represent GDS using standard SummarizedExperiment metaphor. It could save high-throughput genetic/genomic data (e.g., genotyping/sequencing data) within the `SummarizedExperiment` structure. It is more like a container right now. We are working to make it interoperable with the existing functions in `SNPRelate` and `SeqVarTools` designed for gds data. 

## GDSArraySeed, GDSArray, and DelayedArray

GDSArray is the GDS back-end data for DelayedArray, which could be easily stick into SummarizedExperiment, and have on-disk reading and processing of the array data in SE. Common operations on `SummarizedExperiment` could be applied to the assays in GDSArray format, including the subsetting, getters and setters.  
```{r, GDSArray}
seed <- GDSArraySeed(file, "genotype/data")
slotNames(seed)
seed@permute
ga <- GDSArray(file)  
ga <- GDSArray(seed)  ### equivalent
ga[1:5, 1:3, ]
da <- DelayedArray(seed)
### aperm(as.array(ga[1:5, 1:3, ]), c(3,2,1))
```

## SummarizedExperiment with GDS back-end array data.

GDSArray is the GDS back-end DelayedArray, which could stick into the SummarizedExperiment by function `makeSummarizedExperimentFromGDS()`. 
```{r, SEGDSArray}
se <- makeSummarizedExperimentFromGDS(file)
se1 <- makeSummarizedExperimentFromGDS(file, rowDataColumns=c("ID", "REF", "ALT"), colDataColumns=c("family"))
rm(se1)
dim(assay(se))
assay(se)[1:5, 1:3, ]
## saveRDS(se, file="inst/extdata/CEU_Exon_gse.rds")
```


## `save/loadGDSSummarizedExperiment`
Use these 2 functions to save an SE object with ordinary array data into the SE object with GDSArray data, and load the newly generated GDSArray SE object. 

```{r, saveLoad}
### rewrite the DelayedArray data to be an array.
f <- openfn.gds(file)
dat <- read.gdsn(index.gdsn(f, "genotype/data"))
dat <- aperm(dat, perm=c(3,2,1))
closefn.gds(f)
assays(se)[[1]] <- dat
## assays(se) <- list(genotype_data=dat)
asy <- assay(se, 1)
dim(asy)
class(asy)

## save/loadGDSSummarizedExperiment
se1 <- saveGDSSummarizedExperiment(se, dir="my_gds_se", replace=TRUE, verbose=TRUE)
assay(se1, 1)
se2 <- loadGDSSummarizedExperiment(dir="my_gds_se")
assay(se2, 1)
all.equal(assay(se1), assay(se2))
```

## VariantExperiment common SE operations
### Getters and setters
- `rowRanges()` / (`rowData()`), `colData()`, `metadata()`
```{r getSet}
library(SummarizedExperiment)
SummarizedExperiment::rowRanges(se2)
head(SummarizedExperiment::colData(se2))
head(SummarizedExperiment::rowData(se2))
metadata(se2)
```

### subsetting functions
- two-dimension subsetting
```{r, 2d}
se2[1:10, 1:5]
```
- `$` operates on `colData()` columns, for easy sample extraction.
```{r colDataExtraction}
head(SummarizedExperiment::colData(se2))
se2[, se2$family == "1328"]
```
- subsetting by `rowData()` columns.
```{r rowDataExtraction}
head(SummarizedExperiment::rowData(se2))
se2[rowData(se2)$REF == "T",]
```

### Range-based operations

- `subsetByOverlaps()`
`SummarizedExperiment` objects support all of the `findOverlaps()` methods and associated functions.  This includes `subsetByOverlaps()`, which makes it easy to subset a `SummarizedExperiment` object by an interval.

```{r overlap}
subsetByOverlaps(se2, GRanges("22:1-48958933"))
assay(subsetByOverlaps(se2, GRanges("22:1-48958933")))
rowData(subsetByOverlaps(se2, GRanges("22:1-48958933")))
```


# Future work
- Make the meta data for rows and cols on-disk reading. 
- More interoperable with other GDS based functions, including DNA-sequencing data annotation, filteration, and statistical analysis. 
- Develop a pipeline for using `VariantExperiment` as the basic data structure for DNA-sequencing data analysis.  

# Session Info
```{r sessionInfo}
sessionInfo()
```
